repository-data/application/src/main/resources/hcm-config/configuration/update/Migration20180502RevertExternalAttachmentsToEmbedded.groovy

```diff
@@ -29,7 +33,21 @@ class Migration20180502RevertExternalAttachmentsToEmbedded extends BaseNodeUpdat
 
             Node resourceNode = attachmentNode.getNode(NODE_NAME_RESOURCE)
             JcrUtils.ensureIsCheckedOut(resourceNode)
-            resourceNode.setPrimaryType(NODE_TYPE_OLD_RESOURCE)
+
+            // Because the old resource node type doesn't have the relaxed mixin type,
+            // we get an exception here because we have got properties on this node
+            // which are not allowed. To get around this, we suppress the exception and
+            // add the relaxed mixin afterwards manually
+            try {
+                resourceNode.setPrimaryType(NODE_TYPE_OLD_RESOURCE)
+            } catch (ConstraintViolationException cve) {
+                if (cve.getMessage().equals(CONSTRAINT_ERROR_MESSAGE)) {
+                    log.info("Suppressing Constraint error and adding relaxed mixin")
+                    resourceNode.addMixin(HippoStdNodeType.NT_RELAXED)
```

Comment:
Given the exception was thrown when attempting to change the `primaryType` of the node and seeing that the type change is not re-attempted here - does the node type actually get changed despite the exception being thrown or is adding the `relaxed` mixin sufficient instead?

Reply:
Surprisingly, the primary node type change still happens even though the exception is thrown! So we end up changing the primary type and setting the mixin both at the same time when the transaction gets committed.